doc/api/modules.md

```diff
@@ -598,11 +598,26 @@ filename scales linearly with the number of registered extensions.
 In other words, adding extensions slows down the module loader and
 should be discouraged.
 
-#### require.resolve()
+#### require.resolve([options])
 <!-- YAML
 added: v0.3.0
+changes:
+  - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/XXXXX
+    description: The `paths`, `append`, and `prepend` options are now supported.
 -->
 
+* `options` {Object}
+  * `paths` {Array} Paths to resolve module location from. If present, these
+    paths are used instead of the default resolution paths. Note that each of
+    these paths is used as a starting point for the module resolution algorithm,
+    meaning that the `node_modules` hierarchy is checked from this location.
+  * `prepend` {boolean} If `true`, `paths` is prepended to the default
+    resolution paths instead of overwriting them.
+  * `append` {boolean} If `true`, `paths` is appended to the default resolution
+    paths instead of overwriting them. If `append` and `prepend` are both `true`
+    then `prepend` behavior is used.
+
```

Comment:
How about having just `append`, with append on `true`, prepend on `false`, discard on `undefined`?

Reply:
Alternatively, we could go with an approach like 

```js
{
  paths: [ 'these', 'replace', 'the', 'existing', 'path'],
  prepend: [ 'these', 'prepend', 'to', 'the', 'path'],
  append: [ 'these', 'append', 'to', 'the', 'path']
}
```

Reply:
(heh, just saw that @mscdex suggested the same thing above :-) ...)

Reply:
I think that is the suggestion that @mscdex made. I plan to update to that.

Reply:
Is there a way to get the default resolution paths? If yes, `require.resolve` could just accept `options.paths`. Appending/prepending/replacing can be done outside of `require.resolve`

Reply:
`Module._resolveLookupPaths()` gives the paths, but it is not public API. It does seem simplest and most flexible to provide a way to get and set the paths. I could update `resolve()` in this PR to only overwrite the paths with what is passed in, dropping the `append` and `prepend` functionality. Then, add `require.resolve.paths(request)` which returns the default paths for that request.

Reply:
Thinking about it further, that's likely the easiest approach... specifically, have a public API for getting the current resolve path, and an option *only* for overriding it... e.g.

```js
const paths = resolve.getLookupPaths();  // returns a mutable *copy*
paths.unshift('/some/other/path');  // prepend
paths.push('/another/path');   // append
resolve(name, { paths });
```

Reply:
Updating this PR now...